src/encrypt.c

```diff
@@ -0,0 +1,306 @@
+  ece_write_uint32_be(&header.bytes[ECE_KEY_LENGTH], rs);
+  header.bytes[salt->length + 4] = (uint8_t) senderPubKeyLen;
+  if (EC_POINT_point2oct(
+        EC_KEY_get0_group(senderPrivKey), EC_KEY_get0_public_key(senderPrivKey),
+        POINT_CONVERSION_UNCOMPRESSED, &header.bytes[ECE_AES128GCM_HEADER_SIZE],
+        senderPubKeyLen, NULL) != senderPubKeyLen) {
+    err = ECE_ERROR_ENCRYPT;
+    goto end;
+  }
+
+  ece_buf_t ciphertext;
+  ece_buf_slice(payload, header.length, payloadLen, &ciphertext);
+
+  bool isLastRecord = false;
+  size_t blockLen = rs - ECE_AES128GCM_RECORD_OVERHEAD;
+  size_t blockStart = 0;
+  size_t recordStart = 0;
+  size_t counter = 0;
+  while (!isLastRecord) {
+    // Pad so that at least one data byte is in a block.
+    size_t blockPadLen = blockLen - 1;
+    if (blockPadLen > padLen) {
+      blockPadLen = padLen;
+    }
+    padLen -= blockPadLen;
+
+    ece_buf_t block;
+    size_t blockEnd = blockStart + blockLen - blockPadLen;
+    if (blockEnd >= plaintext->length) {
+      blockEnd = plaintext->length;
+      if (!padLen) {
+        // We've reached the last record when the plaintext and padding are
+        // exhausted.
+        isLastRecord = true;
+      }
+    }
+    ece_buf_slice(plaintext, blockStart, blockEnd, &block);
+
+    ece_buf_t blockPad;
+    // `blockPadLen + 1` ensures we always write the delimiter.
+    ece_buf_slice(&pad, 0, blockPadLen + 1, &blockPad);
+    blockPad.bytes[0] = isLastRecord ? 2 : 1;
+
+    ece_buf_t record;
+    size_t recordEnd = recordStart + rs;
+    if (recordEnd >= ciphertext.length) {
+      recordEnd = ciphertext.length;
+    }
+    ece_buf_slice(&ciphertext, recordStart, recordEnd, &record);
+
+    err = ece_aes128gcm_encrypt_block(ctx, &key, &nonce, counter, &block,
+                                      &blockPad, &record);
+    if (err) {
+      goto end;
+    }
+    blockStart = blockEnd;
+    recordStart = recordEnd;
+    counter++;
+  }
+  if (padLen) {
+    err = ECE_ERROR_ENCRYPT_PADDING;
+    goto end;
+  }
+
+end:
+  EVP_CIPHER_CTX_free(ctx);
+  ece_buf_free(&pad);
+  ece_buf_free(&key);
+  ece_buf_free(&nonce);
+  return err;
+}
+
+size_t
+ece_aes128gcm_max_payload_length(uint32_t rs, size_t padLen,
+                                 const ece_buf_t* plaintext) {
+  if (rs <= ECE_AES128GCM_RECORD_OVERHEAD) {
+    return 0;
+  }
+  size_t blockLen = rs - ECE_AES128GCM_RECORD_OVERHEAD;
+  size_t plaintextLen = plaintext->length + padLen;
+  size_t recordCount = (plaintextLen / blockLen) + 1;
+  return ECE_AES128GCM_HEADER_SIZE + ECE_AES128GCM_MAX_KEY_ID_LENGTH +
+         plaintextLen + (ECE_AES128GCM_RECORD_OVERHEAD * recordCount);
+}
+
+int
+ece_aes128gcm_encrypt(const ece_buf_t* rawRecvPubKey,
+                      const ece_buf_t* authSecret, uint32_t rs, size_t padLen,
+                      const ece_buf_t* plaintext, ece_buf_t* payload) {
+  int err = ECE_OK;
+
+  EC_KEY* recvPubKey = NULL;
+  EC_KEY* senderPrivKey = NULL;
```

Comment:
Swap these declarations, for consistency with `ece_aes128gcm_encrypt_with_keys`.